### DISCUSSION TITLE
NEXT.JS redirect not working in getServerSideProps

### DISCUSSION THREAD

**Discussion by lucasmcht** [vercel/next.js#42405] [Help] [answered]
I'm currently developing a web application using NextJS, and in it I've created an authentication system using server-side rendering.

However it sometimes happens that in production (with the next start command) the redirect return of the function is not executed correctly and the user is not redirected to the login page when he is disconnected.

When this bug occurs, refreshing the page fixes the problem.

Here is the function in question

```
export const withAuthServerSideProps = (props?: WithServerSideAuthParams) => {
  return async (context: GetServerSidePropsContext) => {
    const {
      getServerSidePropsFunc,
      redirectCondition,
      requiredRole,
      route = '/signin',
    } = props || {};
    const response = await getUserData(context);
    const isAuthed = !!response?.data;
    let shouldRedirect;
    switch (redirectCondition) {
      case 'isInfluencer':
        shouldRedirect = isAuthed && response.data.role === Roles.INFLUENCER;
        break;
      case 'isAuthed':
        shouldRedirect = isAuthed;
        break;
      case 'isNotAuthed':
        shouldRedirect = !isAuthed;
        break;
      case 'noRedirect':
        shouldRedirect = false;
        break;
    }
    if (
      shouldRedirect ||
      (requiredRole && requiredRole !== response?.data?.role)
    ) {
      return {
        redirect: {
          destination: route,
          permanent: false,
        },
      };
    }
    //et on exécute la fonction du SSR s'il y en a une
    if (getServerSidePropsFunc) {
      try {
        const funcData = await getServerSidePropsFunc(context);
        return {
          props: {
            data: funcData,
            user: response?.data ?? null,
          },
        };
      } catch (e) {
        console.log(e);
        return {
          props: {
            data: null,
            user: response?.data ?? null,
          },
        };
      }
    }
    return {
      props: {
        user: response?.data ?? null,
      },
    };
  };
};
```

And here is how I call it in the app, so that it trigger using SSR

Exemple in a page called **InfluencerPage**

```
export const getServerSideProps = withAuthServerSideProps({
  getServerSidePropsFunc,
});

export default InfluencerPage;
```

The weird thing is that the network response from the service worker is **wrong** (it does not contain redirect)

[![Service Worker response image][1]][1]

But the other network response is **correct**, but the redirect is **not triggered**
[![Network response][2]][2]

  [1]: https://i.stack.imgur.com/U4ArG.png
  [2]: https://i.stack.imgur.com/YlgCq.png

Maybe it's the cache from the service worker that is causing this but i don't know how to solve it

**Comment by icyJoseph** [comment, marked as answer]
Hi,

Yes your service worker is using the incorrect strategy. Refer to this answer https://github.com/vercel/next.js/discussions/42397.

All you gotta do is, if you are using `next-pwa`:

```js
/** @type {import('next').NextConfig} */
const runtimeCaching = require("next-pwa/cache");

const nextDataIndex = runtimeCaching.findIndex(
  (entry) => entry.options.cacheName === "next-data"
);

if (nextDataIndex !== -1) {
  runtimeCaching[nextDataIndex].handler = "NetworkFirst";
} else {
  throw new Error("Failed to find next-data object in runtime caching");
}

const withPWA = require("next-pwa")({
  dest: "public",
  runtimeCaching,
  register: true,
  skipWaiting: true,
  buildExcludes: [/middleware-manifest.json$/],
  disable: process.env.NODE_ENV === 'development'
})

module.exports = withPWA({
  reactStrictMode: false
});
```

Most default PWA setups use `stale-while-revalidate`, but for next-data fetching functions that's very bad because it effectively requires a page reload to replace the cached response.

The workbox strategy that fits best I believe is https://developer.chrome.com/docs/workbox/modules/workbox-strategies/#network-first-network-falling-back-to-cache, network first with fallback to cache.

**Reply by lucasmcht** [reply]
That solved my problem, thanks a lot !
